File: java-devildyw/src/main/java/top/dyw/leetcode/Hot100/L234回文链表.java
package top.dyw.leetcode.Hot100;

public class L234回文链表 {

    public boolean isPalindrome(ListNode head) {
        if (head==null || head.next==null) {
            return true;
        }
        ListNode slow = head, fast = head.next;
        while(fast != null && fast.next != null) {
            fast = fast.next.next;
            slow = slow.next;
        }

        //反转slow链表
        ListNode halfReverseHead = reverseList(slow);
        ListNode cur = head;
        while (cur!=null) {
            if (cur.val != halfReverseHead.val) {
                return false;
            }
            cur = cur.next;
            halfReverseHead = halfReverseHead.next;
        }

        return true;
    }

    public ListNode reverseList(ListNode slow) {
        ListNode head = null;
        while(slow != null) {
            ListNode next = slow.next;
            slow.next = head;
            head = slow;
            slow = next;
        }

        return head;
    }
}

File: golang-ladidol/hot100/middle/L234回文链表.go
package middle

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

// 1、快慢指针找到中点（考虑奇数的长度，需要跳过中间点）
// 2、翻转前半部分
// 3、从中点往两边便利
func isPalindrome(head *ListNode) bool {

	fast := head
	slow := head
	var pre *ListNode

	for fast != nil && fast.Next != nil {
		// 走两步
		fast = fast.Next.Next

		tmp := slow.Next
		slow.Next = pre
		pre = slow
		slow = tmp
	}
	if pre == nil {
		return true
	}

	if fast != nil {
		slow = slow.Next
	}

	// 向前遍历
	for pre != nil && slow != nil {
		if pre.Val != slow.Val {
			return false
		}
		pre = pre.Next
		slow = slow.Next
	}
	return true
}
